tensorflow_probability/python/experimental/mcmc/sample_discarding_kernel_test.py

```diff
@@ -0,0 +1,286 @@
+
+  def test_thinning(self):
+    fake_inner_kernel = TestTransitionKernel()
+    discarder = tfp.experimental.mcmc.SampleDiscardingKernel(
+        fake_inner_kernel,
+        num_steps_between_results=1,)
+    first_state, kernel_results = discarder.one_step(
+        0., discarder.bootstrap_results(0.))
+    second_state, kernel_results = discarder.one_step(
+        first_state, kernel_results)
+    first_state, second_state, kernel_results = self.evaluate([
+        first_state, second_state, kernel_results])
+    self.assertEqual(2, first_state)
+    self.assertEqual(4, second_state)
+    self.assertEqual(2, kernel_results.call_counter)
+    self.assertEqual(4, kernel_results.inner_results.counter_1)
+    self.assertEqual(8, kernel_results.inner_results.counter_2)
+
+  def test_burnin(self):
+    fake_inner_kernel = TestTransitionKernel()
+    discarder = tfp.experimental.mcmc.SampleDiscardingKernel(
+        fake_inner_kernel,
+        num_burnin_steps=5,)
+    sample, kernel_results = discarder.one_step(
+        0., discarder.bootstrap_results(0.))
+    sample, kernel_results = self.evaluate([
+        sample, kernel_results])
+    self.assertEqual(6, sample)
+    self.assertEqual(1, kernel_results.call_counter)
+    self.assertEqual(6, kernel_results.inner_results.counter_1)
+    self.assertEqual(12, kernel_results.inner_results.counter_2)
+
+  def test_no_thinning_or_burnin(self):
+    fake_inner_kernel = TestTransitionKernel()
+    discarder = tfp.experimental.mcmc.SampleDiscardingKernel(
+        fake_inner_kernel,)
+    first_state, kernel_results = discarder.one_step(
+        0., discarder.bootstrap_results(0.))
+    second_state, kernel_results = discarder.one_step(
+        first_state, kernel_results)
+    first_state, second_state, kernel_results = self.evaluate([
+        first_state, second_state, kernel_results])
+    self.assertEqual(1, first_state)
+    self.assertEqual(2, second_state)
+    self.assertEqual(2, kernel_results.call_counter)
+    self.assertEqual(2, kernel_results.inner_results.counter_1)
+    self.assertEqual(4, kernel_results.inner_results.counter_2)
+
+  def test_both_thinning_and_burnin(self):
+    fake_inner_kernel = TestTransitionKernel()
+    discarder = tfp.experimental.mcmc.SampleDiscardingKernel(
+        fake_inner_kernel,
+        num_burnin_steps=10,
+        num_steps_between_results=1,)
+    first_state, kernel_results = discarder.one_step(
+        0., discarder.bootstrap_results(0.))
+    second_state, kernel_results = discarder.one_step(
+        first_state, kernel_results)
+    first_state, second_state, kernel_results = self.evaluate([
+        first_state, second_state, kernel_results])
+    self.assertEqual(12, first_state)
+    self.assertEqual(14, second_state)
+    self.assertEqual(2, kernel_results.call_counter)
+    self.assertEqual(14, kernel_results.inner_results.counter_1)
+    self.assertEqual(28, kernel_results.inner_results.counter_2)
```

Comment:
Please add a quick comment explaining these magic numbers.  Also in at least the next test.